mlir/lib/Dialect/Bufferization/Transforms/BufferResultsToOutParams.cpp

```diff
@@ -43,6 +43,53 @@ static bool hasStaticIdentityLayout(MemRefType type) {
   return type.getLayout().isIdentity();
 }
 
+/// Return the dynamic shapes of the `memref` based on the define op. If the
+/// complete dynamic shape fails to be captured, return an empty value.
+/// Currently, only function parameters are supported for capturing.
+static SmallVector<Value> getDynamicSize(Value memref, func::FuncOp funcOp) {
+  auto *defOp = memref.getDefiningOp();
+  if (!defOp)
+    return {};
+  auto operands = defOp->getOperands();
+  SmallVector<Value> dynamicSizes;
+  for (Value size : operands) {
+    BlockArgument sizeSrc = mlir::dyn_cast<BlockArgument>(size);
+    if (!sizeSrc)
+      return {};
+
+    bool finded = false;
+    for (BlockArgument argument : funcOp.getArguments()) {
+      if (argument == sizeSrc) {
+        dynamicSizes.push_back(argument);
+        finded = true;
+        break;
+      }
+    }
+    if (!finded)
+      return {};
+  }
+  return dynamicSizes;
+}
+
+/// Returns the dynamic sizes at the callee, through the call relationship
+/// between the caller and callee.
+static SmallVector<Value> mapDynamicSizeAtCaller(func::CallOp call,
+                                                 func::FuncOp callee,
+                                                 ValueRange dynamicSizes) {
+  SmallVector<Value> mapedDynamicSizes;
+  for (Value size : dynamicSizes) {
+    auto callOperands = call.getOperands();
+    for (size_t i = 0, e = callOperands.size(); i < e; ++i) {
+      Value src = callOperands[i];
+      BlockArgument dst = callee.getArgument(i);
+      if (size != dst)
+        continue;
+      mapedDynamicSizes.push_back(src);
+    }
+  }
+  return mapedDynamicSizes;
```

Comment:
Add `assert(result.size() == dynamicSizes.size() && "could not find all dynamic sizes")`;